Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm

```diff
@@ -448,6 +514,146 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
     return *m_capabilities;
 }
 
+AVCapturePhotoOutput* AVVideoCaptureSource::photoOutput()
+{
+    assertIsCurrent(RunLoop::main());
+
+    if (!m_photoOutput) {
+        m_photoOutput = adoptNS([PAL::allocAVCapturePhotoOutputInstance() init]);
+
+        if (![session() canAddOutput:m_photoOutput.get()]) {
+            ERROR_LOG_IF(loggerPtr(), LOGIDENTIFIER, "unable to add photo output");
+            return nullptr;
+        }
+        [session() addOutput:m_photoOutput.get()];
+    }
+
+    return m_photoOutput.get();
+}
+
+void AVVideoCaptureSource::resolvePendingPhotoRequest(Vector<uint8_t>&& data, const String& mimeType)
+{
+    Locker lock { m_photoLock };
+
+    if (!m_photoProducer)
+        return;
+
+    m_photoProducer->resolve(std::make_pair(WTFMove(data), mimeType));
+    m_photoProducer = nullptr;
+}
+
+void AVVideoCaptureSource::rejectPendingPhotoRequest(const String& error)
+{
+    Locker lock { m_photoLock };
+
+    if (!m_photoProducer)
+        return;
+
+    m_photoProducer->reject(error);
+    m_photoProducer = nullptr;
+}
+
+IntSize AVVideoCaptureSource::maxPhotoSizeForCurrentPreset(IntSize requestedSize) const
+{
+    ASSERT(isMainThread());
+
+    CMVideoDimensions bestMaxPhotoSize;
+
+    auto *format = [m_device activeFormat];
+    if ([format respondsToSelector:@selector(supportedMaxPhotoDimensions)]) {
+        NSArray<NSValue*> *maxPhotoDimensions = format.supportedMaxPhotoDimensions;
+        if (!maxPhotoDimensions.count)
+            return { };
+
+        bestMaxPhotoSize = maxPhotoDimensions.firstObject.CMVideoDimensionsValue;
+        for (NSValue *value in maxPhotoDimensions) {
+            CMVideoDimensions dimensions = value.CMVideoDimensionsValue;
+            if (dimensions.width >= requestedSize.width() && dimensions.height >= requestedSize.height()) {
+                if (dimensions.width * dimensions.height < bestMaxPhotoSize.width * bestMaxPhotoSize.height)
+                    bestMaxPhotoSize = dimensions;
+            }
+        }
+    } else {
+        if (!m_currentPreset)
+            return { };
+
+        return m_currentPreset->size();
+    }
+
+    return { bestMaxPhotoSize.width, bestMaxPhotoSize.height };
+}
+
+RetainPtr<AVCapturePhotoSettings> AVVideoCaptureSource::photoConfiguration(const PhotoSettings& photoSettings)
+{
+    assertIsCurrent(RunLoop::main());
+
+    IntSize requestedPhotoDimensions = { 0, 0 };
+    if (photoSettings.imageHeight && photoSettings.imageWidth)
+        requestedPhotoDimensions = { static_cast<int>(*photoSettings.imageWidth), static_cast<int>(*photoSettings.imageHeight) };
+
+    AVCapturePhotoSettings* avPhotoSettings = [PAL::getAVCapturePhotoSettingsClass() photoSettingsWithFormat:@{
+        AVVideoCodecKey : AVVideoCodecTypeJPEG,
+        AVVideoCompressionPropertiesKey : @{ AVVideoQualityKey : @(1) }
+    }];
+
+#if PLATFORM(IOS_FAMILY)
+    auto* photoOutput = this->photoOutput();
+    ASSERT(photoOutput);
```

Comment:
If canAddOutput return false, this fails. I would return with nullptr and reject the promise in such a case.